src/main/java/org/opensearch/security/resources/ResourceSharingIndexListener.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.security.resources;
+
+import java.io.IOException;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import org.opensearch.accesscontrol.resources.CreatedBy;
+import org.opensearch.accesscontrol.resources.ResourceSharing;
+import org.opensearch.client.Client;
+import org.opensearch.core.index.shard.ShardId;
+import org.opensearch.index.engine.Engine;
+import org.opensearch.index.shard.IndexingOperationListener;
+import org.opensearch.security.support.ConfigConstants;
+import org.opensearch.security.user.User;
+import org.opensearch.threadpool.ThreadPool;
+
+/**
+ * This class implements an index operation listener for operations performed on resources stored in plugin's indices
+ * These indices are defined on bootstrap and configured to listen in OpenSearchSecurityPlugin.java
+ */
+public class ResourceSharingIndexListener implements IndexingOperationListener {
+
+    private final static Logger log = LogManager.getLogger(ResourceSharingIndexListener.class);
+
+    private static final ResourceSharingIndexListener INSTANCE = new ResourceSharingIndexListener();
+    private ResourceSharingIndexHandler resourceSharingIndexHandler;
+
+    private boolean initialized;
```

Comment:
Can we not use this extra boolean and just base it on the indexlistener itself?

Reply:
Its to ensure that ReosurceSharingIndexListener is only initialized once.